Add tutorial for Cloudflare Workers with Deno

jebreuer opened the pull request:
See title.

As discussed earlier today in the Telegram group here is what I found worked for me.
Let me know if anything needs to be fixed. I tried to follow the contribution guide to the best of my understanding.

I based my writing on what I found in cloudflare-workers.md. Thanks to the author. <3

Kind regards

KnorpelSenf commented:
The relevant discussion in the chat started at https://t.me/grammyjs/122790

KnorpelSenf reviewed (changes requested):
Thank you so much!

Ideally, instead of having a tutorial for Service Workers with Node and separate one for Module Workers with Deno, we'd only have one tutorial for Module Workers which explains how to set things up for either Deno or Node. We do not need a tutorial for Service Workers. Do you think this is possible? :)

jebreuer commented:
Hi all!

I've pushed a few suggested changes.

I would like to disagree on the opinion that both approaches are that similar so they could be merged into "one page". What they have in common is that both use "CloudFlare Workers" and GrammY and that's it.
I just tried to setup things the Nodejs way as described [here](https://developers.cloudflare.com/workers/get-started/guide/) and I actually failed. From my perspective both approaches are different enough to give each of them an isolated space as to not confuse the newbie developer. The Nodejs approach contains so much boilerplate that the Deno approach would just get lost between the Nodejs noise.

I'd like to leave it to those who actually want to use the Nodejs approach to prove me wrong.

If someone comes up with a working Nodejs example I'll happily give it a second look again and try to merge things. Just let me know. For now I consider that out of scope for my PR.

Kind regards
Jens

KnorpelSenf commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,124 @@
+interface Environment {
+  BOT_TOKEN: string;
+}
+
+export default {
+  fetch(request: Request, env: Environment) {
+    try {
+      const bot = new Bot(env.BOT_TOKEN);
```
Recreating the bot instance per request is problematic because the initialisation will have to happen again and again. Given that CFW has quite an ugly API with the env vars not being accessible during module loading, we can work around this by storing the bot info in a variable as soon as it's known.

```suggestion
let botInfo: UserFromGetMe | undefined = undefined
export default {
  fetch(request: Request, env: Environment) {
    try {
      const bot = new Bot(env.BOT_TOKEN, { botInfo });
      if (botInfo === undefined) {
        await bot.init()
        botInfo = bot.botInfo
      }
```

Please see if this type-checks, coded on GitHub.

jebreuer replied:
Done.

KnorpelSenf commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,124 @@
+
+      const cb = webhookCallback(bot, "cloudflare-mod");
+
+      return cb(request);
```
Errors won't be caught, right?
```suggestion
      return await cb(request);
```

jebreuer replied:
Done.

KnorpelSenf reviewed (changes requested):
> I would like to disagree on the opinion that both approaches are that similar so they could be merged into "one page". What they have in common is that both use "CloudFlare Workers" and GrammY and that's it.
I just tried to setup things the Nodejs way as described [here](https://developers.cloudflare.com/workers/get-started/guide/) and I actually failed. From my perspective both approaches are different enough to give each of them an isolated space as to not confuse the newbie developer. The Nodejs approach contains so much boilerplate that the Deno approach would just get lost between the Nodejs noise.

Alright, then let's leave it separate for now. If we find a good way to merge the guides, we can do this in a subsequent pull request.

I have a few smaller comments, but otherwise this LGTM:

CikiMomogi commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,124 @@
+
+Create a new directory and create a new file `.denoflare` in that directory.
+Put the following contents in the file:
+
```
```suggestion

> Note: The "$schema" key in the following JSON code specifies a fixed version in its URL ("v0.5.11").
> At the time of writing, this was the latest version available.
> You should update them to the [newest version](https://github.com/skymethod/denoflare/releases).

```

jebreuer replied:
Done.

CikiMomogi commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,124 @@
+[Cloudflare Workers](https://workers.cloudflare.com/) is a public serverless computing platform that offers a convenient and simple solution for running small workloads at the [edge](https://en.wikipedia.org/wiki/Edge_computing).
+
+This guide will take you through the process of hosting your Telegram bot on Cloudflare Workers.
+
```
Don't forget to apply the same changes to the [other doc](https://grammy.dev/hosting/cloudflare-workers.html) as well.
```suggestion

:::tip Looking for the Node.js version?
This tutorial explains how to deploy a Telegram bot to Cloudflare Workers using Deno.
If you're looking for the Node.js version, please check out [this tutorial](./cloudflare-workers.md) instead.
:::

```

jebreuer replied:
Done.

jebreuer commented on site/docs/.vuepress/configs/en.ts:
```diff
@@ -361,6 +361,10 @@ export const localeEn: LocaleConfig<DefaultThemeLocaleData> = {
                 text: "Cloudflare Workers",
```
Done.

jebreuer commented:
Hi all!

I implemented/added the requested changes and cleaned up the branch a bit.

Please have a look if it works for you now.

Kind regards
Jens

CikiMomogi commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,141 @@
+
+Make sure you have [Deno](https://deno.land/) and [Denoflare](https://denoflare.dev/) installed.
+
+Create a new directory and create a new file `.denoflare` in that directory.
```
```suggestion
Create a new directory, and create a new file `.denoflare` in that directory.
```

CikiMomogi commented on site/docs/hosting/cloudflare-workers-deno.md:
````diff
@@ -0,0 +1,141 @@
+}
+```
+
+Make sure to replace `YOUR_ACCOUNT_ID`, `YOUR_API_TOKEN` and `YOUR_BOT_TOKEN` appropriately.
+When creating your API token you can chose the `Edit Cloudflare Workers` preset from the pre-configured permissions.
````
```suggestion
Make sure to replace `YOUR_ACCOUNT_ID`, `YOUR_API_TOKEN`, and `YOUR_BOT_TOKEN` appropriately.
When creating your API token, you can choose the `Edit Cloudflare Workers` preset from the pre-configured permissions.
```

CikiMomogi commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,141 @@
+
+> Note: The "$schema" key in the following JSON code specifies a fixed version in its URL ("v0.5.11").
+> At the time of writing, this was the latest version available.
+> You should update them to the [newest version](https://github.com/skymethod/denoflare/releases)
```
```suggestion
> You should update them to the [newest version](https://github.com/skymethod/denoflare/releases).
```

CikiMomogi reviewed (approved):
LGTM. Just a few minor improvements below.

jebreuer commented:
Added requested changes, squashed and pushed.

LGTM? ;-)

jebreuer commented:
Updated Co-Authors.

jebreuer commented:
Friendly reminder, @KnorpelSenf. ;)

roj1512 commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,141 @@
+
+[Cloudflare Workers](https://workers.cloudflare.com/) is a public serverless computing platform that offers a convenient and simple solution for running small workloads at the [edge](https://en.wikipedia.org/wiki/Edge_computing).
+
+This guide will take you through the process of hosting your Telegram bot on Cloudflare Workers.
```
```suggestion
This guide will take you through the process of hosting your bot on Cloudflare Workers.
```

roj1512 commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,141 @@
+:::tip Looking for the Node.js Version?
+This tutorial explains how to deploy a Telegram bot to Cloudflare Workers using Deno.
+If you're looking for the Node.js version, please check out [this tutorial](./cloudflare-workers.md) instead.
+:::
```
```suggestion

:::
```

roj1512 commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,141 @@
+
+## Prerequisites
+
+To follow along, please make sure that you have a [Cloudflare account](https://dash.cloudflare.com/login) with your workers subdomain [configured](https://dash.cloudflare.com/?account=workers).
```
```suggestion
To follow along, please make sure that you have a [Cloudflare account](https://dash.cloudflare.com/login) with your Workers subdomain [configured](https://dash.cloudflare.com/?account=workers).
```

roj1512 commented on site/docs/hosting/cloudflare-workers.md:
```diff
@@ -1,11 +1,16 @@
 This guide will take you through the process of hosting your Telegram bots on Cloudflare Workers.
 
+:::tip Looking for the Deno Version?
+This tutorial explains how to deploy a Telegram bot to Cloudflare Workers using Node.js.
```
```suggestion

This tutorial explains how to deploy a Telegram bot to Cloudflare Workers using Node.js.
```

roj1512 commented on site/docs/hosting/cloudflare-workers.md:
```diff
@@ -1,11 +1,16 @@
 
 This guide will take you through the process of hosting your Telegram bots on Cloudflare Workers.
 
+:::tip Looking for the Deno Version?
+This tutorial explains how to deploy a Telegram bot to Cloudflare Workers using Node.js.
+If you're looking for the Deno version, please check out [this tutorial](./cloudflare-workers-deno.md) instead.
+:::
```
```suggestion
:::tip Looking for the Deno Version?

This tutorial explains how to deploy a Telegram bot to Cloudflare Workers using Node.js.
If you're looking for the Deno version, please check out [this tutorial](./cloudflare-workers-deno.md) instead.

:::
```

roj1512 commented on site/docs/hosting/cloudflare-workers.md:
```diff
@@ -1,11 +1,16 @@
 
 This guide will take you through the process of hosting your Telegram bots on Cloudflare Workers.
 
+:::tip Looking for the Deno Version?
```
```suggestion
:::tip Looking for the Deno Version?

```

CikiMomogi commented on site/docs/hosting/cloudflare-workers.md:
```diff
@@ -1,11 +1,16 @@
-# Hosting: Cloudflare Workers
+# Hosting: Cloudflare Workers (Node)
```
```suggestion
# Hosting: Cloudflare Workers (Node.js)
```

CikiMomogi commented on site/docs/.vuepress/configs/en.ts:
```diff
@@ -358,9 +358,13 @@ export const localeEn: LocaleConfig<DefaultThemeLocaleData> = {
               },
               {
-                text: "Cloudflare Workers",
+                text: "Cloudflare Workers (Node)",
```
```suggestion
                text: "Cloudflare Workers (Node.js)",
```

KnorpelSenf replied:
Perhaps we should use the old link to refer to the Deno guide, and then make a new page for Node.js? Deno tooling is far superior, so it makes sense to guide people there.

KnorpelSenf commented:
> Sorry for the delay, will leave a review in the coming days!

Whoopsie, that was optimistic. I did not forget about this, though. I'll come back to it tomorrow when I review #766, too.

KnorpelSenf commented on site/docs/hosting/cloudflare-workers-deno.md:
```diff
@@ -0,0 +1,141 @@
+Create a new directory, and create a new file `.denoflare` in that directory.
+Put the following contents in the file:
+
+> Note: The "$schema" key in the following JSON code specifies a fixed version in its URL ("v0.5.11").
```
```suggestion
> Note: The "$schema" key in the following JSON code specifies a fixed version in its URL ("v0.5.12").
```
(artefacts of my late review)

KnorpelSenf commented on site/docs/hosting/cloudflare-workers-deno.md:
````diff
@@ -0,0 +1,141 @@
+
+```json
+{
+  "$schema": "https://raw.githubusercontent.com/skymethod/denoflare/v0.5.11/common/config.schema.json",
````
```suggestion
  "$schema": "https://raw.githubusercontent.com/skymethod/denoflare/v0.5.12/common/config.schema.json",
```

KnorpelSenf reviewed (changes requested):
Just a few details, this is really neat!

KnorpelSenf commented:
@roj1512 perhaps you can take it from here, the content is great that way IMO. We can do our style nitpicking, but otherwise I think there's nothing else to do here

roj1512 commented:
I can’t push changes.

KnorpelSenf commented:
We want to merge this into a temporary branch for translation anyway. Might as well do it now.

KnorpelSenf commented:
@roj1512 are you going to do that, or are you instead planning on waiting for @jebreuer to fix things up?

roj1512 commented:
Thanks for making me remember.
